go/libraries/doltcore/sqle/statspro/listener.go

```diff
@@ -0,0 +1,268 @@
+// Copyright 2025 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statspro
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/dolthub/go-mysql-server/sql"
+
+	"github.com/dolthub/dolt/go/libraries/doltcore/dbfactory"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle"
+	"github.com/dolthub/dolt/go/libraries/doltcore/sqle/dsess"
+)
+
+var ErrStatsIssuerPaused = fmt.Errorf("stats issuer is paused")
+
+type listenerEvent uint16
+
+const (
+	leUnknown               = listenerEvent(iota)
+	leSwap    listenerEvent = 1 << 0
+	leStop    listenerEvent = 1 << 1
+	leGc      listenerEvent = 1 << 2
+	leFlush   listenerEvent = 1 << 3
+)
+
+func (sc *StatsController) signalListener(s listenerEvent) {
+	var root, keep *listenMsg
+	n := sc.listeners
+	for n != nil {
+		if (n.e|leStop)&s > 0 {
+			n.c <- s
+			close(n.c)
+		} else if root == nil {
+			root = n
+			keep = n
+		} else {
+			keep.n = n
+			keep = n
+		}
+		n = n.n
```

Comment:
Urge to kill... rising...

Reply:
When you're iterating over a linked list, the variable names `curr` and `next` are customary

Reply:
![image](https://github.com/user-attachments/assets/49ef95f6-7725-4899-8632-ca0257c79a2b)